authentication-via-social-networks/README.md

```diff
@@ -0,0 +1,763 @@
+
+describe AuthVerificationPolicy do
+  let(:auth) { double(:omniauth, provider: provider) }
+
+  describe ".verified?" do
+    subject { described_class.new(auth).verified? }
+
+    context "when provider is Facebook" do
+      let(:provider) { "facebook" }
+
+      before do
+        allow(auth).to receive_message_chain(:info, :verified?).and_return(true)
+        allow(auth).to receive_message_chain(:extra, :raw_info, :verified?).and_return(true)
+      end
+
+      it "returns corresponding value" do
+        expect(subject).to eq(true)
+      end
+    end
+
+    context "when provider is Google" do
+      let(:provider) { "google_oauth2" }
+
+      before do
+        allow(auth).to receive_message_chain(:extra, :raw_info, :email_verified?).and_return(true)
+      end
+
+      it "returns corresponding value" do
+        expect(subject).to eq(true)
+      end
+    end
+
+    context "when provider checking is not defined" do
+      let(:provider) { "another" }
+
+      it "raises Exception" do
+        expect { subject }
+          .to raise_error(AuthVerificationPolicy::OauthError, I18n.t("omniauth.verification.not_implemented", kind: provider))
+      end
+    end
+  end
+end
+```
+
+```ruby
+# spec/rails_helper.rb
+...
+OmniAuth.config.test_mode = true
+...
+```
+
+```ruby
+# spec/support/matchers/have_connected_account.rb
+RSpec::Matchers.define :have_connected_account do |identity|
+  match do
+    within ".js-identities" do
+      have_text(identity)
+    end
+  end
+end
+```
+
+```ruby
+# spec/support/shared_examples/omniauth_stub.rb
+require "rails_helper"
+
+shared_context :auth_hashie do
+  let(:auth_hashie) do
+    OmniAuth::AuthHash.new(
+      provider: "facebook",
+      uid: "123545",
+      info: {
+        email: "joe@bloggs.com",
+        name: "Joe Bloggs",
+        verified: true
+      },
+      extra: {
+        raw_info: {
+          email: "joe@bloggs.com",
+          name: "Joe Bloggs",
+          verified: true,
+          email_verified: true
+        }
+      }
+    )
+  end
+end
+
+shared_context :stub_omniauth do
+  background do
+    OmniAuth.config.mock_auth[:facebook] = OmniAuth::AuthHash.new(
+      provider: "facebook",
+      uid: "123545",
+      info: {
+        email: "joe@bloggs.com",
+        name: "Joe Bloggs",
+        verified: true
+      },
+      extra: {
+        raw_info: {
+          email: "joe@bloggs.com",
+          name: "Joe Bloggs",
+          verified: true,
+          email_verified: true
+        }
+      }
+    )
+  end
+end
+
+shared_context :stub_not_verified_omniauth do
+  background do
+    OmniAuth.config.mock_auth[:facebook] = OmniAuth::AuthHash.new(
+      provider: "facebook",
+      uid: "123545",
+      info: {
+        email: "joe@bloggs.com",
+        name: "Joe Bloggs",
+        verified: false
+      },
+      extra: {
+        raw_info: {
+          email: "joe@bloggs.com",
+          name: "Joe Bloggs",
+          verified: false,
+          email_verified: false
```

Comment:
just wonder, can we use _create(:auth_hashie)_ here aswell, but with some kind of deep merging '_verified_' keys. I recall there was something like _Hash#dig_ which might can be used for this